src/pages/index.js

```diff
@@ -38,54 +89,113 @@ const createCard = (cardData) => {
 const addCardObject = {
   popupSelector: "#add-card-modal",
   handleFormSubmit: (inputValues) => {
-    const cardElement = createCard(inputValues);
-    addCardForm.popupForm.reset();
-    addCardFormValidator.toggleButtonState();
-    cardSelection.addItem(cardElement);
-    addCardForm.close();
+    api
+      .addCard(inputValues.name, inputValues.link)
+      .then((res) => {
+        // cardElement = createCard(inputValues);
+        // const cardElement = createCard(inputValues);
+        const cardElement = createCard(res);
+        // addCardForm.popupForm.reset();
+        // addCardFormValidator.toggleButtonState();
+        cardSelection.addItem(cardElement);
+        addCardForm.close();
+      })
+      .finally(() => {
+        addCardForm.popupForm.reset();
+        addCardFormValidator.toggleButtonState();
+      })
+      .catch((res) => {
+        console.log(`There is an error in the program: ${res}`);
+      });
   },
 };
 //the edit profile object
 const editProfileObject = {
   popupSelector: "#edit-modal",
   handleFormSubmit: (inputValues) => {
-    userProfile.setUserInfo(inputValues.profileName, inputValues.profileJob);
-    editProfileForm.popupForm.reset();
-    editProfileFormValidator.toggleButtonState();
-    editProfileForm.close();
+    editProfileForm.renderLoading("true", "Saving...");
```

Comment:
**Could be improved**

If it’s interesting for you here is how we can make a universal function for handling any submit. We can get rid of such duplicating as **loading effect**, **closing** and **catching errors**


```js

// You can make a universal function that accepts a request, popup instance and optional loading text  
function handleSubmit(request, popupInstance, loadingText = "Saving...") {

  // here we change the button text
  popupInstance.renderLoading(true, loadingText);
  request()
    .then(() => {
      // We need to close only in `then`
      popupInstance.close()
    })
      // we need to catch possible errors
      // console.error is used to handle errors if you don’t have any other ways for that
     .catch(console.error)
    // in `finally` we need to return the initial button text back in any case
    .finally(() => {
      popupInstance.renderLoading(false);
    });
}



```


You need to pass the **request** into the call of `handleSubmit` like that:



```js

// here is an example of the profile form handling
function handleProfileFormSubmit(inputValues) {
  // we create a function that returns a promise
  function makeRequest() {
    // `return` lets us use a promise chain `then, catch, finally` inside `handleSubmit`
    return api.editProfile(inputValues).then((userData) => {
      userInfo.setUserInfo(userData)
    });
  }
  // Here we call the function passing the request, popup instance and if we need some other loading text we can pass it as the 3rd argument
  handleSubmit(makeRequest, profilePopup);
}



```